### ISSUE TITLE
Unable to build library and examples

### ISSUE THREAD

**Issue by eamreyes** [gpgme-sharp/gpgme-sharp#17]
I'm running inside a WSL2 Debian instance with:

```
~/gpgme-sharp$ dotnet --list-sdks
6.0.415 [/usr/share/dotnet/sdk]

~/gpgme-sharp$ dotnet --list-runtimes
Microsoft.AspNetCore.App 6.0.23 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
Microsoft.NETCore.App 6.0.23 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

```
Trying to build without modification I encounter the [The "ILRepack" task has been declared or used incorrectly, or failed during construction. Check the spelling of the task name and the assembly name.](https://github.com/peters/ILRepack.MSBuild.Task/issues/28)

If I update the ILRepack package I get 
```

  ILRepack: Adding assembly for merge: /home/default/gpgme-sharp/gpgme-sharp/bin/Debug/net40/GPGME.Native.Win32.dll
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error : Specified method is not supported.
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error :    at Mono.Cecil.ModuleParameters.GetCurrentRuntime()
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error :    at Mono.Cecil.ModuleParameters..ctor()
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error :    at ILRepacking.ILRepack.Repack()
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error :    at ILRepack.MSBuild.Task.ILRepack.Execute()
  ILRepack: Adding assembly for merge: /home/default/gpgme-sharp/gpgme-sharp/bin/Debug/netcoreapp3.1/GPGME.Native.Shared.dll
  ILRepack: Adding assembly for merge: /home/default/gpgme-sharp/gpgme-sharp/bin/Debug/netcoreapp3.1/GPGME.Native.Unix.dll
  ILRepack: Adding assembly for merge: /home/default/gpgme-sharp/gpgme-sharp/bin/Debug/netcoreapp3.1/GPGME.Native.Win32.dll
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error : Specified method is not supported.
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error :    at Mono.Cecil.ModuleParameters.GetCurrentRuntime()
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error :    at Mono.Cecil.ModuleParameters..ctor()
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error :    at ILRepacking.ILRepack.Repack()
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error :    at ILRepack.MSBuild.Task.ILRepack.Execute()

Build FAILED.

/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error : Specified method is not supported.
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error :    at Mono.Cecil.ModuleParameters.GetCurrentRuntime()
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error :    at Mono.Cecil.ModuleParameters..ctor()
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error :    at ILRepacking.ILRepack.Repack()
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error :    at ILRepack.MSBuild.Task.ILRepack.Execute()
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error : Specified method is not supported.
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error :    at Mono.Cecil.ModuleParameters.GetCurrentRuntime()
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error :    at Mono.Cecil.ModuleParameters..ctor()
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error :    at ILRepacking.ILRepack.Repack()
/home/default/gpgme-sharp/gpgme-sharp/gpgme-sharp.csproj(48,5): error :    at ILRepack.MSBuild.Task.ILRepack.Execute()
    0 Warning(s)
    2 Error(s)

```

I'm running `dotnet build` in the Example project folders.

**Comment by Daniel15** [comment]
I want to get rid of the ILPack since it's not needed any more that the minimum .NET version has been bumped to .NET Core 3.1. It was only needed as a hack due to the fact that the GPGME library has a different name on Windows compared to Linux (see https://github.com/gpgme-sharp/gpgme-sharp/pull/7). In .NET Core 3.1 and above, we can use a [custom importer resolver](https://learn.microsoft.com/en-us/dotnet/standard/native-interop/native-library-loading) to achieve the same thing in a non-hacky way.

I'll try to do that soon 😄

**Comment by Daniel15** [comment]
Also I've only ever tested the build on Windows, so maybe there's some issues with it on other OSes. (the resulting library works fine cross-platform though)

**Comment by eamreyes** [comment]
I've been playing with it on Linux, removed ILPack and added missing references to the Example projects. I get this result.

```
$ dotnet run
Unhandled exception. System.BadImageFormatException: Could not load file or assembly '/home/default/gpgme-sharp/Examples/CreatePgpKey/bin/Debug/net6/CreatePgpKey.dll'. An attempt was made to load a program with an incorrect format.

File name: '/home/default/gpgme-sharp/Examples/CreatePgpKey/bin/Debug/net6/CreatePgpKey.dll'
```

I'll look at the custom resolver. I'm curious on how these cross-platform libraries are put together.

**Comment by Daniel15** [comment]
@eamreyes Looking through my repo, it turns out I did most of the work for this (removing the hacks and ILRepack) about 1.5 years ago 😅 I just created a PR #18 that you can try.

For the "An attempt was made to load a program with an incorrect format" error, that's something I need to address too. You need to remove this from the `.csproj` of the example you're trying to run:
```
<PlatformTarget>x86</PlatformTarget>
```

On Windows, the GPGME library is 32-bit, so the entire app needs to run as a 32-bit app. However, on Linux, the library is 64-bit if you're using a 64-bit distro. I need to figure out how to force 32-bit on Windows but allow either 32- or 64-bit on Linux.

I've split that out into a separate issue #19

**Comment by eamreyes** [comment]
While playing with the custom resolver, I found a 64 bit version of the library on Windows. 

@Daniel15 `C:\Program Files (x86)\Gpg4win\bin_64\libgpgme6-11`

It works with the example projects, but I'm not sure if it's intended for consumption.

**Comment by Daniel15** [comment]
@eamreyes Wow, that's a good catch. I didn't see that. I usually don't expect to find 64-bit apps in `Program Files (x86)` 😛  I'll modify the custom resolver to explicitly look in that directory when running a 64-bit build on Windows.

**Comment by Daniel15** [comment]
@eamreyes  ~~I was looking into this again, and it looks like the 64-bit version is a lot older? At least on my system, `bin_64\libgpgme-11.dll`  is version 1.16.1.17795 dated 2021-12-31, whereas `bin\libgpgme-11.dll` is version 1.21.0.23539 dated 2023-07-14. The version in `C:\Program Files (x86)\GnuPG\bin` is yet another version - `1.20.0.17694` dated `2023-07-04`.~~

In `bin_64`, there's also a DLL called `libgpgme6-11.dll` with the same 1.21.0.23539 version as the 32-bit version. I don't know what the 6 means in the name, and there's only one result in Google for it.

🤔🤔🤔🤔

Edit: It looks like `libgpgme-11.dll` in the `bin_64` was left over from an old installation, and new versions only include `libgpgme6-11.dll`. I guess the `6` means 64-bit?

**Comment by eamreyes** [comment]
I'm not sure.  We may need to dig into [the community forums](https://www.gpg4win.org/community.html)

**Comment by Daniel15** [comment]
@eamreyes  Can you please try again with the latest version of the repo? It should handle both 32-bit and 64-bit apps now. On Windows, it'll look in the `bin_64` directory for 64-bit apps and the regular `bin` directory for 32-bit apps. On Linux it just uses the default library resolver since it should work fine for both.

I'll publish the library to NuGet as a beta soon too.

**Comment by Daniel15** [comment]
New version released as 3.0.0-beta1, with 64-bit support on Windows.
